toolkit/livedata/src/main/kotlin/br/com/arch/toolkit/livedata/response/SwapResponseLiveData.kt

```diff
@@ -38,31 +38,47 @@ class SwapResponseLiveData<T> : ResponseLiveData<T>() {
         }
         lastSource = source
     }
-
     /**
      * Changes the actual DataSource, with transformation
      *
      * @param source The ResponseLiveData to replicate the value
+     * @param transformAsync Indicate swapSource will execute synchronously or asynchronously
      * @param transformation Receives the data of the source and change to T value
      *
      * @see SwapResponseLiveData.swapSource
      */
-    fun <R> swapSource(source: ResponseLiveData<R>, transformation: (R) -> T) {
+    fun <R> swapSource(
+            source: ResponseLiveData<R>,
+            transformAsync: Boolean,
+            transformation: (R) -> T
+    ) {
         clearSource()
         sourceLiveData.addSource(source) { data ->
-            async {
-                if (data == null) return@async
-                val newValue = when (data.status) {
-                    SUCCESS -> DataResult(data.data?.let(transformation), null, SUCCESS)
-                    ERROR -> DataResult<T>(null, data.error, ERROR)
-                    LOADING -> DataResult<T>(null, null, LOADING)
-                }
-                if (value != newValue) postValue(newValue)
+            if (transformAsync) {
+                async { doTransformation(data, transformation) }
+            } else {
+                doTransformation(data, transformation)
             }
         }
         lastSource = source
     }
 
+    fun <R> swapSource(
+            source: ResponseLiveData<R>,
+            transformation: (R) -> T
+    ) {
+        swapSource(source, false, transformation)
+    }
+
+    private fun <R> doTransformation(data: DataResult<R>, transformation: (R) -> T) {
+        val newValue = when (data.status) {
+            SUCCESS -> DataResult<T>(data.data?.let(transformation), null, SUCCESS)
+            ERROR -> DataResult<T>(null, data.error, ERROR)
+            LOADING -> DataResult<T>(null, null, LOADING)
+        }
+        if (value != newValue) value = newValue
```

Comment:
when the transformation runs async, the setValue method will crash because... Use postValue is necessary when it is running out of the main thread, I think we can add a check "if is in main thread, user setValue, use postValue otherwise", what do you think?

Reply:
It makes absolute sense, I've just updated the method with this check.